### PR TITLE
Fix false negatives for css-in-js object notation in color-hex-case

### PR DESCRIPTION
**Pull request by Dru89**
As discussed in #4826, `style-search` has false positives when used with `css-in-js` object notation. This change updates the `color-hex-case` rule to use `postcss-value-parser` instead, which has support for `css-in-js` as a syntax.

The unit tests now also include a suite for testing both `upper` and `lower` options for `color-hex-case` with the `css-in-js` syntax.

### CODE REVIEW

**Comment by Dru89** [comment]
This is my first contribution, so let me know if there's something I missed with respect to formatting or code style.

File: lib/rules/color-hex-case/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/index.js:39]: Let's pull this out into a self-documenting function, e.g.:

```js
if (isIgnoredFunction(node)) return false;

// at the bottom of file
function isIgnoredFunction({type, value}) {
  return type === 'function' && IGNORED_FUNCTIONS.has(value.toLowerCase())
}
```

The comment is redundant then.

File: lib/rules/color-hex-case/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/index.js:44]: Let's do the same for this:

```js
if (!isHexColor(node)) return;
```

File: lib/rules/color-hex-case/__tests__/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/__tests__/index.js:175]: Let's ident all these template literals for readability, e.g.:

```js
                   {
			code: `
                              import styled from 'styled-components';
                              export const s = styled.a({
	                             color: "pink",
                              });
                         `,
```

File: lib/rules/color-hex-case/__tests__/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/__tests__/index.js:178]:
```suggestion
	stroke: "url(#AAA)",
```

Let's use a string that looks like an uppercase hex to make sure it's ignored.

File: lib/rules/color-hex-case/__tests__/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/__tests__/index.js:187]:
```suggestion
	color: "PINK",
```

Let's use uppercase to make sure it's ignored.

File: lib/rules/color-hex-case/__tests__/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/__tests__/index.js:195]:
```suggestion
	color: "#aaa",
```

Let's use lowercase, and let's move this test first as it's the clearest example of a valid pattern.

File: lib/rules/color-hex-case/__tests__/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/__tests__/index.js:237]:
```suggestion
	content: '"#ABABAB"',
```

Let's use a valid hex.

File: lib/rules/color-hex-case/__tests__/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/__tests__/index.js:226]: I think we can delete this test as `padding: 000` isn't valid CSS value.

File: lib/rules/color-hex-case/__tests__/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/__tests__/index.js:200]: Let's combine the next three tests into valid CSS code, rather than use a `something` property, e.g.:

```css
background: linear-gradient(#aaa, #ffff, #0000ffcc);
```

This will cover the 4 and 8 digit tests.

File: lib/rules/color-hex-case/__tests__/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/__tests__/index.js:256]:
```suggestion
	color: "#aBABAA",
```

File: lib/rules/color-hex-case/__tests__/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/__tests__/index.js:262]:
```suggestion
	color: "#ababaa",
```

File: lib/rules/color-hex-case/__tests__/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/__tests__/index.js:266]:
```suggestion
			message: messages.expected('#aBABAA', '#ababaa'),
```

File: lib/rules/color-hex-case/__tests__/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/__tests__/index.js:176]: I believe can remove `import styled from 'styled-components';` from all the tests as we're explicitly specifying the syntax.

File: lib/rules/color-hex-case/__tests__/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/__tests__/index.js:271]: Let's combine the next three tests into one valid example:

```
background: linear-gradient(#AAA, #FaFa #0000FFcc);
```

File: lib/rules/color-hex-case/__tests__/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/__tests__/index.js:328]: Let's apply all the requested changes above to the `upper` tests too.

**Review by jeddy3 — CHANGES_REQUESTED** [review]
@Dru89 Thanks! It's looking great so far, and so much cleaner now that the value parser is being used.

I've made a few requested changes.

**Comment by Dru89** [comment]
Thanks for taking the time to look at all of this @jeddy3!

I've gone through and made the changes you requested. For many of them (like the invalid five hex colors, e.g. `#aBABA`), I pulled those directly from the test cases above. (My assumption was that the original tests still wanted to make sure that "invalid" hex colors still got transformed.)

Should we transform the original tests to also use more valid CSS? Or just leave them as they were before?

File: lib/rules/color-hex-case/__tests__/index.js
Comment (jeddy3) [inline lib/rules/color-hex-case/__tests__/index.js:168]: I think we got our wires crossed with the indents. Let's indent the template literals as normal blocks, like so:

```js
testRule({
	ruleName,
	config: ['upper'],
	syntax: 'css-in-js',
	fix: true,
	accept: [
		{
			code: `
				export const s = styled.a({
				    color: "#AAA",
				});
			`,
		},
	],

	reject: [
		{
			code: `
				export const s = styled.a({
				    color: "#aBABAB",
				});
			`,
			fixed: `
				export const s = styled.a({
					color: "#ABABAB",
				});
			`,
			message: messages.expected('#aBABAB', '#ABABAB'),
			line: 3,
			column: 15,
		},
	],
});
```

**Review by jeddy3 — CHANGES_REQUESTED** [review]
Fantastic, thanks for making the changes! It's almost there. Just one last tweak with the indents.

> Should we transform the original tests to also use more valid CSS? Or just leave them as they were before?

Let's leave them as they are. We default to using valid CSS, but deviating in this case makes sense as it's testing something particular. Whereas in the `testRule` for `"css-in-js"` were testing that objects are read correctly, rather than the edge cases of the rule itself, and so valid CSS is appropriate here.

File: lib/rules/color-hex-case/__tests__/index.js
Comment (Dru89) [reply lib/rules/color-hex-case/__tests__/index.js:168]: Ah, apologies! Github's 8-spaces-per-tab made it look like you were trying to align it with the backtick, not just trying to indent it by one line. I'll take care of this today.

**Review by jeddy3 — APPROVED** [review]
@Dru89 Thanks. Looks great to me!

It's awesome to see the improvements with the refactoring to the value parser.

**Comment by Dru89** [comment]
Thanks! I was gonna start in on some of the other `style-search` use cases like `hex-color-length`, `color-no-hex`, and `value-no-vendor-prefix`.

Should I wait until this one gets the second approval and merged? Also, would you prefer one PR per rule?

**Comment by jeddy3** [comment]
> Thanks! I was gonna start in on some of the other style-search use cases...

Fantastic!

> Should I wait until this one gets the second approval and merged? Also, would you prefer one PR per rule?

No need to wait. Yes, one pull request per rule, please.

File: lib/rules/color-hex-case/index.js
Comment (hudochenkov) [inline lib/rules/color-hex-case/index.js:18]: Shall we use RegExp, which looks for valid HEX only?

```suggestion
const HEX = /^#([\da-f]{3,4}|[\da-f]{6}|[\da-f]{8})$/i;
```

https://regexr.com/5k70k

CleanShot 2021-01-13 at 22 49 53

File: lib/rules/color-hex-case/index.js
Comment (hudochenkov) [inline lib/rules/color-hex-case/index.js:86]: Return is not used anywhere.

File: lib/rules/color-hex-case/index.js
Comment (Dru89) [reply lib/rules/color-hex-case/index.js:18]: My goal was to match the previous style search (and not break existing tests) for this one, and the original query was:

```
/^#[0-9A-Za-z]+/
```

But if we're ok with a potentially "breaking" change that will only look at valid hex colors, I'm ok to make that.

Comment (hudochenkov) [reply lib/rules/color-hex-case/index.js:18]: I would say it's not a breaking change, but a bugfix. What do you think, @jeddy3?

Comment (jeddy3) [reply lib/rules/color-hex-case/index.js:18]: I believe the rule was originally designed to check the case of anything that _sort of looks like a hex colour_, i.e. a word that starts with hash and is followed by numbers and letters, as:

1. people may want to come up with their own hex-like languages extensions
2. we already had the [color-no-invalid-hex rule](https://stylelint.io/user-guide/rules/color-no-invalid-hex) rule the people could use to check for hex colour validity against the specs

As it's more extensible, and because users may already be relying on this behaviour, I suggest we keep it as.

File: lib/rules/color-hex-case/index.js
Comment (jeddy3) [reply lib/rules/color-hex-case/index.js:86]: @Dru89 It looks like this is the only change we need to make for a second approval.

Comment (jeddy3) [reply lib/rules/color-hex-case/index.js:86]: @Dru89 It looks like this is the only change we need to make for a second approval.

**Comment by Dru89** [comment]
Sorry for the delay in getting this updated. Last week was a bit crazy, but I've addressed that last piece of feedback.

**Review by hudochenkov — APPROVED** [review]
👍

**Comment by jeddy3** [comment]
Changelog:

> - Fixed: `color-hex-case` false negatives for css-in-js object notation ([#5101](https://github.com/stylelint/stylelint/pull/5101)).

@Dru89 Thanks again for your first two pull request. Excellent job!

**Pull request merged**
